src/NuGet.Core/NuGet.Packaging/PackageReaderBase.cs

```diff
@@ -536,6 +536,46 @@ protected static string GetNuspecFile(IEnumerable<string> files)
             return nuspecPaths.Single();
         }
 
+        /// <summary>
+        /// Validate file entry in package is not traversed outside of the expected extraction path.
+        /// Eg: file entry like ../../foo.dll can get outside of the expected extraction path.
+        /// </summary>
+        protected static void ValidatePackageEntry(string normalizedDestination, string normalizedFilePath, PackageIdentity packageIdentity)
+        {
+            // Destination and filePath must be normalized.
+            var fullPath = Path.GetFullPath(Path.Combine(normalizedDestination, normalizedFilePath));
+
+            if(!fullPath.StartsWith(normalizedDestination, StringComparison.Ordinal) || fullPath.Length == normalizedDestination.Length)
+            {
+                throw new UnsafePackageEntryException(string.Format(
```

Comment:
Where is this caught for `PluginPackageReader` ?

Reply:
don't need to catch it here, just let it throw

Reply:
the outside caller, it can be UI or PMC or command.